#!/usr/bin/env python3
"""
Configuration file for ANNOVAR-fast database mappings and settings
"""

import os

# Database path â€” all databases (tabix .txt.gz, mRNA .fa) live here
HUMANDB_DIR = "/data/alvin/annovar/humandb-tbi"

# Database mapping from annovar-hg38.awk protocol line
# Operation types: 'g' = gene, 'r' = region, 'f' = filter
# Format for each:
#   1000g: special 6-col format (chr, pos, ref, obs, AF, rsid) - no chr prefix
#   generic_filter: chr, start, end, ref, obs, data...
#   snp141: UCSC snp format with bin column
#   region: various UCSC formats with bin column
#   gene: refGene/ensGene format
DATABASE_CONFIG = {
    # Gene annotations (operation 'g')
    'refGene': {
        'file': 'hg38_refGene.txt.gz',
        'type': 'gene',
        'operation': 'g',
        'columns': ['Func.refGene', 'Gene.refGene', 'GeneDetail.refGene', 'ExonicFunc.refGene', 'AAChange.refGene'],
        'splicing_threshold': 100,
    },
    'ensGene': {
        'file': 'hg38_ensGene_new.txt.gz',
        'type': 'gene',
        'operation': 'g',
        'columns': ['Func.ensGene', 'Gene.ensGene', 'GeneDetail.ensGene', 'ExonicFunc.ensGene', 'AAChange.ensGene'],
        'splicing_threshold': 2,
    },

    # Region annotations (operation 'r')
    'cytoBand': {
        'file': 'hg38_cytoBand.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['cytoBand'],
        'has_bin': False,
        'positionCols': (0, 1, 2),
        'scoreCols': (),
        'colsToOutput': (0, 3),
        'has_chr_prefix': True,
    },
    'genomicSuperDups': {
        'file': 'hg38_genomicSuperDups.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['genomicSuperDups'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (27, 27),
        'colsToOutput': (4,),
        'has_chr_prefix': True,
    },
    'targetScanS': {
        'file': 'hg38_targetScanS.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['targetScanS'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (5, 5),
        'colsToOutput': (4,),
        'has_chr_prefix': True,
    },
    'cosmic': {
        'file': 'hg38_cosmic.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['cosmic'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (),
        'colsToOutput': (4,),
        'has_chr_prefix': False,
    },
    'gwasCatalog': {
        'file': 'hg38_gwasCatalog.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['gwasCatalog'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (),
        'colsToOutput': (10,),
        'has_chr_prefix': True,
    },
    'encRegTfbsClustered': {
        'file': 'hg38_encRegTfbsClustered.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['encRegTfbsClustered'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (),
        'colsToOutput': (4,),
        'has_chr_prefix': True,
    },
    'wgEncodeRegDnaseClustered': {
        'file': 'hg38_wgEncodeRegDnaseClustered.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['wgEncodeRegDnaseClustered'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (5,),
        'colsToOutput': (4,),
        'has_chr_prefix': True,
    },
    'wgRna': {
        'file': 'hg38_wgRna.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['wgRna'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (5, 5),
        'colsToOutput': (4,),
        'has_chr_prefix': True,
    },
    'dgvMerged': {
        'file': 'hg38_dgvMerged.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['dgvMerged'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (),
        'colsToOutput': (4,),
        'has_chr_prefix': True,
    },
    'phastConsElements30way': {
        'file': 'hg38_phastConsElements30way.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['phastConsElements30way'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (4, 5),
        'colsToOutput': (4,),
        'has_chr_prefix': True,
    },
    'phastConsElements100way': {
        'file': 'hg38_phastConsElements100way.txt.gz',
        'type': 'region',
        'operation': 'r',
        'columns': ['phastConsElements100way'],
        'has_bin': True,
        'positionCols': (1, 2, 3),
        'scoreCols': (4, 5),
        'colsToOutput': (4,),
        'has_chr_prefix': True,
    },

    # Filter annotations (operation 'f')
    'esp6500siv2_all': {
        'file': 'hg38_esp6500siv2_all.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['esp6500siv2_all'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'exac03': {
        'file': 'hg38_exac03.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['ExAC_ALL', 'ExAC_AFR', 'ExAC_AMR', 'ExAC_EAS', 'ExAC_FIN', 'ExAC_NFE', 'ExAC_OTH', 'ExAC_SAS'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    '1000g2015aug_all': {
        'file': 'hg38_ALL.sites.2015_08.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': '1000g',
        'columns': ['1000g2015aug_all'],
        'has_chr_prefix': False,
    },
    '1000g2015aug_afr': {
        'file': 'hg38_AFR.sites.2015_08.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': '1000g',
        'columns': ['1000g2015aug_afr'],
        'has_chr_prefix': False,
    },
    '1000g2015aug_eas': {
        'file': 'hg38_EAS.sites.2015_08.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': '1000g',
        'columns': ['1000g2015aug_eas'],
        'has_chr_prefix': False,
    },
    '1000g2015aug_amr': {
        'file': 'hg38_AMR.sites.2015_08.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': '1000g',
        'columns': ['1000g2015aug_amr'],
        'has_chr_prefix': False,
    },
    '1000g2015aug_eur': {
        'file': 'hg38_EUR.sites.2015_08.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': '1000g',
        'columns': ['1000g2015aug_eur'],
        'has_chr_prefix': False,
    },
    '1000g2015aug_sas': {
        'file': 'hg38_SAS.sites.2015_08.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': '1000g',
        'columns': ['1000g2015aug_sas'],
        'has_chr_prefix': False,
    },
    'intervar_20250721': {
        'file': 'hg38_intervar_20250721.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['InterVar_automated', 'PVS1', 'PS1', 'PS2', 'PS3', 'PS4', 'PM1', 'PM2', 'PM3', 'PM4', 'PM5', 'PM6', 'PP1', 'PP2', 'PP3', 'PP4', 'PP5', 'BA1', 'BS1', 'BS2', 'BS3', 'BS4', 'BP1', 'BP2', 'BP3', 'BP4', 'BP5', 'BP6', 'BP7'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'kaviar_20150923': {
        'file': 'hg38_kaviar_20150923.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['Kaviar_AF', 'Kaviar_AC', 'Kaviar_AN'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'mcap': {
        'file': 'hg38_mcap.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['MCAP'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'revel': {
        'file': 'hg38_revel.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['REVEL'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'gnomad41_exome': {
        'file': 'hg38_gnomad41_exome.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['gnomad41_exome_AF', 'gnomad41_exome_AF_grpmax', 'gnomad41_exome_AF_XY', 'gnomad41_exome_AF_XX', 'gnomad41_exome_AF_raw', 'gnomad41_exome_AF_afr', 'gnomad41_exome_AF_sas', 'gnomad41_exome_AF_amr', 'gnomad41_exome_AF_eas', 'gnomad41_exome_AF_nfe', 'gnomad41_exome_AF_fin', 'gnomad41_exome_AF_asj', 'gnomad41_exome_AF_remaining', 'gnomad41_exome_faf95', 'gnomad41_exome_faf99', 'gnomad41_exome_fafmax_faf95_max', 'gnomad41_exome_AF_grpmax'],
        'data_start_col': 5,
        'has_chr_prefix': True,
    },
    'gnomad41_genome': {
        'file': 'hg38_gnomad41_genome.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['gnomad41_genome_AF', 'gnomad41_genome_AF_raw', 'gnomad41_genome_AF_XY', 'gnomad41_genome_AF_XX', 'gnomad41_genome_AF_afr', 'gnomad41_genome_AF_ami', 'gnomad41_genome_AF_amr', 'gnomad41_genome_AF_asj', 'gnomad41_genome_AF_eas', 'gnomad41_genome_AF_fin', 'gnomad41_genome_AF_nfe', 'gnomad41_genome_AF_remaining', 'gnomad41_genome_AF_sas'],
        'data_start_col': 5,
        'has_chr_prefix': True,
    },
    'snp141': {
        'file': 'hg38_snp141.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'snp',
        'columns': ['snp141'],
        'has_chr_prefix': True,
    },
    'ljb26_all': {
        'file': 'hg38_ljb26_all.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['SIFT_score', 'SIFT_pred', 'Polyphen2_HDIV_score', 'Polyphen2_HDIV_pred', 'Polyphen2_HVAR_score', 'Polyphen2_HVAR_pred', 'LRT_score', 'LRT_pred', 'MutationTaster_score', 'MutationTaster_pred', 'MutationAssessor_score', 'MutationAssessor_pred', 'FATHMM_score', 'FATHMM_pred', 'MetaSVM_score', 'MetaSVM_pred', 'MetaLR_score', 'MetaLR_pred', 'VEST4_score', 'CADD_raw', 'CADD_phred', 'GERP++_RS', 'phyloP30way_mammalian', 'phyloP100way_vertebrate', 'SiPhy_29way_logOdds'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'clinvar_20250721': {
        'file': 'hg38_clinvar_20250721.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['CLNALLELEID', 'CLNDN', 'CLNDISDB', 'CLNREVSTAT', 'CLNSIG'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'dbscsnv11': {
        'file': 'hg38_dbscsnv11.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['dbscSNV_ADA_SCORE', 'dbscSNV_RF_SCORE'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'hrcr1': {
        'file': 'hg38_hrcr1.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['HRC_AF', 'HRC_AC', 'HRC_AN', 'HRC_non1000G_AF', 'HRC_non1000G_AC', 'HRC_non1000G_AN'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'gme': {
        'file': 'hg38_gme.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['GME_AF', 'GME_NWA', 'GME_NEA', 'GME_AP', 'GME_Israel', 'GME_SD', 'GME_TP', 'GME_CA'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'icgc28': {
        'file': 'hg38_icgc28.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['ICGC_Id', 'ICGC_Occurrence'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'TCGA': {
        'file': 'hg38_TCGA.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['TCGA'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'civic': {
        'file': 'hg38_civic.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['civic'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'tumorportal': {
        'file': 'hg38_tumorportal.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['tumorportal'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'cg69': {
        'file': 'hg38_cg69.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['cg69'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
    'nci60': {
        'file': 'hg38_nci60.txt.gz',
        'type': 'filter',
        'operation': 'f',
        'format': 'generic',
        'columns': ['nci60'],
        'data_start_col': 5,
        'has_chr_prefix': False,
    },
}

# Database processing order (matches the --protocol order in annovar-hg38.awk)
DATABASE_ORDER = [
    'refGene', 'ensGene',
    'cytoBand', 'genomicSuperDups',
    'esp6500siv2_all', 'exac03',
    '1000g2015aug_all', '1000g2015aug_afr', '1000g2015aug_eas', '1000g2015aug_amr', '1000g2015aug_eur', '1000g2015aug_sas',
    'intervar_20250721', 'kaviar_20150923', 'mcap', 'revel',
    'gnomad41_exome', 'gnomad41_genome',
    'snp141', 'ljb26_all', 'clinvar_20250721',
    'cosmic',
    'dbscsnv11', 'hrcr1', 'gme', 'icgc28',
    'TCGA', 'civic', 'tumorportal',
    'gwasCatalog', 'encRegTfbsClustered', 'wgEncodeRegDnaseClustered',
    'cg69', 'nci60', 'wgRna', 'dgvMerged',
    'phastConsElements30way', 'phastConsElements100way', 'targetScanS',
]

# All output columns in order (from original ANNOVAR output)
OUTPUT_COLUMNS = [
    'Chr', 'Start', 'End', 'Ref', 'Alt',
    'Func.refGene', 'Gene.refGene', 'GeneDetail.refGene', 'ExonicFunc.refGene', 'AAChange.refGene',
    'Func.ensGene', 'Gene.ensGene', 'GeneDetail.ensGene', 'ExonicFunc.ensGene', 'AAChange.ensGene',
    'cytoBand', 'genomicSuperDups',
    'esp6500siv2_all',
    'ExAC_ALL', 'ExAC_AFR', 'ExAC_AMR', 'ExAC_EAS', 'ExAC_FIN', 'ExAC_NFE', 'ExAC_OTH', 'ExAC_SAS',
    '1000g2015aug_all', '1000g2015aug_afr', '1000g2015aug_eas', '1000g2015aug_amr', '1000g2015aug_eur', '1000g2015aug_sas',
    'InterVar_automated', 'PVS1', 'PS1', 'PS2', 'PS3', 'PS4', 'PM1', 'PM2', 'PM3', 'PM4', 'PM5', 'PM6', 'PP1', 'PP2', 'PP3', 'PP4', 'PP5', 'BA1', 'BS1', 'BS2', 'BS3', 'BS4', 'BP1', 'BP2', 'BP3', 'BP4', 'BP5', 'BP6', 'BP7',
    'Kaviar_AF', 'Kaviar_AC', 'Kaviar_AN',
    'MCAP', 'REVEL',
    'gnomad41_exome_AF', 'gnomad41_exome_AF_grpmax', 'gnomad41_exome_AF_XY', 'gnomad41_exome_AF_XX', 'gnomad41_exome_AF_raw', 'gnomad41_exome_AF_afr', 'gnomad41_exome_AF_sas', 'gnomad41_exome_AF_amr', 'gnomad41_exome_AF_eas', 'gnomad41_exome_AF_nfe', 'gnomad41_exome_AF_fin', 'gnomad41_exome_AF_asj', 'gnomad41_exome_AF_remaining', 'gnomad41_exome_faf95', 'gnomad41_exome_faf99', 'gnomad41_exome_fafmax_faf95_max', 'gnomad41_exome_AF_grpmax',
    'gnomad41_genome_AF', 'gnomad41_genome_AF_raw', 'gnomad41_genome_AF_XY', 'gnomad41_genome_AF_XX', 'gnomad41_genome_AF_afr', 'gnomad41_genome_AF_ami', 'gnomad41_genome_AF_amr', 'gnomad41_genome_AF_asj', 'gnomad41_genome_AF_eas', 'gnomad41_genome_AF_fin', 'gnomad41_genome_AF_nfe', 'gnomad41_genome_AF_remaining', 'gnomad41_genome_AF_sas',
    'snp141', 'SIFT_score', 'SIFT_pred', 'Polyphen2_HDIV_score', 'Polyphen2_HDIV_pred', 'Polyphen2_HVAR_score', 'Polyphen2_HVAR_pred', 'LRT_score', 'LRT_pred', 'MutationTaster_score', 'MutationTaster_pred', 'MutationAssessor_score', 'MutationAssessor_pred', 'FATHMM_score', 'FATHMM_pred', 'MetaSVM_score', 'MetaSVM_pred', 'MetaLR_score', 'MetaLR_pred', 'VEST4_score', 'CADD_raw', 'CADD_phred', 'GERP++_RS', 'phyloP30way_mammalian', 'phyloP100way_vertebrate', 'SiPhy_29way_logOdds',
    'CLNALLELEID', 'CLNDN', 'CLNDISDB', 'CLNREVSTAT', 'CLNSIG',
    'cosmic', 'dbscSNV_ADA_SCORE', 'dbscSNV_RF_SCORE',
    'HRC_AF', 'HRC_AC', 'HRC_AN', 'HRC_non1000G_AF', 'HRC_non1000G_AC', 'HRC_non1000G_AN',
    'GME_AF', 'GME_NWA', 'GME_NEA', 'GME_AP', 'GME_Israel', 'GME_SD', 'GME_TP', 'GME_CA',
    'ICGC_Id', 'ICGC_Occurrence',
    'TCGA', 'civic', 'tumorportal', 'gwasCatalog',
    'encRegTfbsClustered', 'wgEncodeRegDnaseClustered', 'cg69', 'nci60', 'wgRna', 'dgvMerged',
    'phastConsElements30way', 'phastConsElements100way', 'targetScanS',
    'Otherinfo1', 'Otherinfo2', 'Otherinfo3', 'Otherinfo4', 'Otherinfo5', 'Otherinfo6', 'Otherinfo7', 'Otherinfo8', 'Otherinfo9', 'Otherinfo10', 'Otherinfo11'
]

# Performance settings
MAX_WORKERS = os.cpu_count() or 8
CACHE_SIZE = 100000
BATCH_SIZE = 1000

# NA string (from original ANNOVAR --nastring .)
NA_STRING = '.'

# Near-gene distance for intergenic classification
NEARGENE_DISTANCE = 1000
